lib/src/clojure_lsp/dep_graph.clj

```diff
@@ -0,0 +1,344 @@
+    {:internal? false
+     :langs #{:clj :cljs}
+     :namespaces #{clojure.tools.cli}}})
+
+;;;; DB Maintenance
+
+(def ^:private empty-multiset {})
+
+(defn ^:private mulitset-conj [ms x]
+  (update ms x (fnil inc 0)))
+
+(defn ^:private multiset-disj [ms x]
+  (if-let [v (get ms x)]
+    (let [new-v (dec v)]
+      (if (zero? new-v)
+        (dissoc ms x)
+        (assoc ms x new-v)))
+    ms))
+
+(defn ms-distinct [ms] (some-> ms keys))
+
+(defn ms-overlaps-set? [ms s]
+  ;; Conceptually set overlap is symmetric, but with our implementation of
+  ;; multiset, in practice we must check whether the items of the set are in the
+  ;; multiset, not vice-versa.
+  (boolean (and ms (some ms s))))
+
+(def ^:private ms-conj (fnil mulitset-conj empty-multiset))
+(def ^:private ms-disj (fnil multiset-disj empty-multiset))
+(def ^:private s-conj (fnil conj #{}))
+(def ^:private s-disj (fnil disj #{}))
+
+(defn file-internal? [db filename]
+  (get-in db [:file-meta filename :internal?]))
+
+(defn ^:private update-usage [db f {:keys [from name alias filename] :as element}]
+  (let [from-file (get-in db [:file-meta filename])
+        from-langs (or (some-> element :lang list set) (:langs from-file))]
+    ;; A dep-graph item is a summary of the ways a namespace is used across the
+    ;; whole project. We keep multisets of most of its data, so that we know not
+    ;; only that a namespace `aaa` is aliased as `a`, but that it is aliased as
+    ;; `a` 7 times. If the alias is removed from a single file we decrement to
+    ;; 6, but we don't lose track of the fact that it's still aliased as `a`
+    ;; somewhere else.
+    (-> db
+        (update-in [:dep-graph from :dependencies] f name)
+        (update-in [:dep-graph name :dependents] f from)
+        (update-in [:dep-graph name :aliases] f alias)
+        ;; NOTE: We could store :from-filenames, and look up whether any of them
+        ;; are internal. But this way keeps that lookup out of q/ns-aliases,
+        ;; which is on the hotpath in completion.
+        ;; TODO: is it wrong that once from-internal?, always from-internal?
+        (update-in [:dep-graph name :from-internal?] #(or % (:internal? from-file)))
+        ;; NOTE: :from-langs is used only in add-missing-libspec, so it's not on
+        ;; a hotpath, but :from-internal? has already established this pattern.
+        (update-in [:dep-graph name :from-langs] #(reduce f % from-langs)))))
+
+(defn ^:private update-usages [db f namespace-usages]
+  (reduce #(update-usage %1 f %2) db namespace-usages))
+
+(defn ^:private remove-usages [db usages] (update-usages db ms-disj usages))
+(defn ^:private add-usages [db usages]    (update-usages db ms-conj usages))
+
+(defn ^:private update-definition [db f {:keys [name filename]}]
+  (let [in-file (get-in db [:file-meta filename])]
+    (-> db
+        (update-in [:dep-graph name :files] f filename)
+        ;; TODO: is it wrong that once an internal ns, always an internal ns?
```

Comment:
I think we can assume it's always internal, I can't see any case where this is not true

Reply:
Yeah, it'd be very unusual for a namespace to switch from internal to external. The only way it would happen is if you re-opened a dependency's namespace momentarily (external -> internal), then removed the code that did that (internal -> external). It's such a weird scenario that I'll just remove the comment.